src/security/Handshake.cc

```diff
@@ -504,6 +521,78 @@ Security::HandshakeParser::parseSniExtension(const SBuf &extensionData) const
     return SBuf(); // SNI extension lacks host_name
 }
 
+/// RFC 8446 Section 4.2.1: SupportedVersions extension
+void
+Security::HandshakeParser::parseSupportedVersionsExtension(const SBuf &extensionData) const
+{
+    // Upon detecting a quoted RFC MUST violation, this parser immediately
+    // returns, ignoring the entire extension and resulting in Squid relying on
+    // the legacy_version field value or another (valid) supported_versions
+    // extension. The alternative would be to reject the whole handshake as
+    // invalid. Deployment experience will show which alternative is the best.
+
+    // Please note that several of these MUSTs also imply certain likely
+    // handling of a hypothetical next TLS version (e.g., v1.4).
+
+    // RFC 8446 Section 4.1.2:
+    // In TLS 1.3, the client indicates its version preferences in the
+    // "supported_versions" extension (Section 4.2.1) and the legacy_version
+    // field MUST be set to 0x0303, which is the version number for TLS 1.2.
+    //
+    // RFC 8446 Section 4.2.1:
+    // A server which negotiates TLS 1.3 MUST respond by sending a
+    // "supported_versions" extension containing the selected version value
+    // (0x0304).  It MUST set the ServerHello.legacy_version field to 0x0303
+    // (TLS 1.2).
+    //
+    // Ignore supported_versions senders violating legacy_version MUSTs above:
+    if (details->tlsSupportedVersion != AnyP::ProtocolVersion(AnyP::PROTO_TLS, 1, 2))
+        return;
+
+    AnyP::ProtocolVersion supportedVersionMax;
+    if (messageSource == fromClient) {
+        Parser::BinaryTokenizer tkList(extensionData);
+        Parser::BinaryTokenizer tkVersions(tkList.pstring8("SupportedVersions"));
+        while (!tkVersions.atEnd()) {
+            const auto version = ParseProtocolVersion(tkVersions, "supported_version");
+            if (!supportedVersionMax || TlsVersionEarlierThan(supportedVersionMax, version))
+                supportedVersionMax = version;
+        }
+
+        // ignore empty supported_versions
+        if (!supportedVersionMax)
+            return;
+
+        // supportedVersionMax here may be "earlier" than tlsSupportedVersion: A
+        // TLS v1.3 client may try to negotiate a _legacy_ version X with a TLS
+        // v1.3 server by sending supported_versions containing just X.
+    } else {
+        assert(messageSource == fromServer);
+        Parser::BinaryTokenizer tkVersion(extensionData, "selected_version");
+        const auto version = ParseProtocolVersion(tkVersion);
+        // RFC 8446 Section 4.2.1:
+        // A server which negotiates a version of TLS prior to TLS 1.3 [...]
+        // MUST NOT send the "supported_versions" extension.
+        if (Tls1p2orEarlier(version))
```

Comment:
Given the local static member from above this can be simplified as:
```suggestion
        if (!(version > Tls1p2))
```

Reply:
The operators *can* be used when appropriate. The care here is only needed because of the involvement of any PROTO_SSL as a desired true condition. Thus the combining of "!" rather than simpler "<=" which is also available.

I saw the Must checks and think they are not necessary for this particular condition. Protocols other than TLS (if possible to occur, maybe NONE does) can/should safely just return leaving the TLS/1.3 extension logic ignored. And this if-statement is exactly a choice whether to abort parsing of the extension with a return.


As for commit 0a0fb9c - I do not think using a static inside an inline function is a good idea. That can result in multiple static variables being declared across many build objects, which can cause conflicts with the symbol linking and initialization.

Reply:
> The operators can be used when appropriate.

Agreed, but I do not think this is one of those appropriate cases. There is a high risk that a developer looking at `a > b` code will write `x < y` code elsewhere, thinking that this is a regular, mathematically correct comparison operation that can be safely inverted, negated, etc. for all operands. Due to the SSL-TLS barrier, the raw comparison operation is _not_ mathematically correct. We should avoid it as much as possible. Future code safety is the primary reason why, after evaluating several options, we added wrappers like Tls1p2orEarlier().

> Must checks [...] are not necessary for this particular condition.

We should not entrap developers with an example of an innocently-looking seemingly-easy-to-reuse code that will silently fail to work as expected in many other seemingly-similar contexts.

I agree with Christos that this code should continue to use wrappers. @yadij, please reconsider your request to remove this well-considered safety feature, especially on the "this particular code works without it" grounds.

BTW, if somebody wants to investigate whether the SSL-TLS barrier itself should be removed (by removing `PROTO_SSL` and mapping its versions to non-existent `PROTO_TLS` versions), I would welcome that research. If, as the result of that complicated project, the barrier is removed, the safety wrappers should be removed as well. Or we can just wait for SSL to disappear from any real traffic we care about.

> As for commit 0a0fb9c - I do not think using a static inside an inline function is a good idea.

You are right. I have undone that overreaction in 3403e0d. Please see another discussion for the reasons behind that reversal: https://github.com/squid-cache/squid/pull/588#discussion_r412357158

Reply:
> >    As for commit 0a0fb9c - I do not think using a static inside an inline function is a good idea.

> You are right. 

Just a comment on this. I had the same fear before use static inside inline function so I  test it before using it. There is not any problem at least using g++ and clang compilers.

Moreover I am copying from the (draft of) c++98 standard  I have.  Section 7.1.2 says:
_"... A static local variable in an extern inline function always refers to the same object. A string literal in an extern inline function is the same object in different translation units. "_

I suppose the same policy is used for c++11 and later.

Reply:
@chtsanti, please note that my "you are right" remark referred to the "bad idea" _conclusion_ only. The reasoning behind that conclusion was problematic, but I did not quote or comment on that reasoning in hope to reduce points of conflict -- it is the conclusion that matters in this case :-).

Your own conclusion is also correct -- statics inside `extern` inlined functions do not multiply across translation units. Such multiplication would make most of them practically unusable! The expensive mutex-checking code dealing with static variable guards does go into every caller of an inlined function, but the static variable itself is stored outside the caller.